### PR TITLE
Automatically minimize failing queries

### PR DESCRIPTION
**Pull request by systay**
## Description

Adds test minimizer functionality. This is a technique borrowed from Haskells QuickCheck - for a given failing query, we want to simplify the query to the bare essentials that still produce the same error. 
The idea is to use this infrastructure either in conjunction with fuzzing, or when a bug is reported, we can use this infrastructure to minimize the query that fails.

This PR adds a test method that can be used when planner bugs are reported: `TestSimplifyBuggyQuery`
It also hooks up the evalengine fuzzer to the minimizer, so if the fuzzer finds problems, we should get a simpler expression to work on.

### CODE REVIEW

**Review by frouioui — APPROVED** [review]
I like the idea behind this pull request, thank you. I left a few comments/questions, looks awesome overall!

File: go/vt/vtgate/simplifier/expression_simplifier.go [hunk 1]
```diff
@@ -0,0 +1,200 @@
...
+		if s.queue == nil {
+			return nil
+		}
+	case *sqlparser.ColName:
+		// we can try to replace the column with a literal value
+		s.queue = []sqlparser.Expr{sqlparser.NewIntLiteral("0")}
+	default:
+		return nil
+	}
+	return s.Next()
```
[the remark this answers is not in the document]
Comment (systay) [reply go/vt/vtgate/simplifier/expression_simplifier.go:199]: I think this logic is the way it should be. 

`return nil` means that there is no way to make the expression simpler, and that is the default for unknown types.

`return s.Next()` means that we have added stuff to the queue and should start emptying the queue.

[the remark this answers is not in the document]
Comment (systay) [reply go/vt/vtgate/simplifier/expression_simplifier.go]: updated the code and made it easier to follow

[the remark this answers is not in the document]
Comment (frouioui) [reply go/vt/vtgate/simplifier/expression_simplifier.go]: got it, 2006f42 covers this

**Review by frouioui — APPROVED** [review]
Still looks good to me after the latest changes

File: go/vt/vtgate/planbuilder/abstract/join.go [hunk 2]
```diff
@@ -45,8 +45,8 @@ func (j *Join) PushPredicate(expr sqlparser.Expr, semTable *semantics.SemTable)
 		// we are looking for predicates like `tbl.col = <>` or `<> = tbl.col`,
 		// where tbl is on the rhs of the left outer join
 		if cmp, isCmp := expr.(*sqlparser.ComparisonExpr); isCmp && cmp.Operator != sqlparser.NullSafeEqualOp &&
-			sqlparser.IsColName(cmp.Left) && semTable.RecursiveDeps(cmp.Left).IsSolvedBy(j.RHS.TableID()) ||
-			sqlparser.IsColName(cmp.Right) && semTable.RecursiveDeps(cmp.Right).IsSolvedBy(j.RHS.TableID()) {
+			(sqlparser.IsColName(cmp.Left) && semTable.RecursiveDeps(cmp.Left).IsSolvedBy(j.RHS.TableID()) ||
+				sqlparser.IsColName(cmp.Right) && semTable.RecursiveDeps(cmp.Right).IsSolvedBy(j.RHS.TableID())) {
```
[the remark this answers is not in the document]
Comment (systay) [reply go/vt/vtgate/planbuilder/abstract/join.go:49]: :heavy_check_mark:

**Review by harshit-gangal — APPROVED** [review]

File: go/vt/vtgate/planbuilder/testdata/unsupported_cases.txt [hunk 3]
```diff
@@ -59,7 +59,7 @@ Gen4 plan same as above
 Gen4 plan same as above
 
 # left join with expressions
-"select user.id, user_extra.col+1 from user left join user_extra on user.col = user_extra.col"
```
[the remark this answers is not in the document]
Comment (harshit-gangal) [reply go/vt/vtgate/planbuilder/testdata/unsupported_cases.txt:62]: the simplified query also fails, got that. :)

**Pull request merged**